### ISSUE TITLE
Multiple HTTP auth dialogs are spawned at the same time

### ISSUE THREAD

**Issue by 14mRh4X0r** [passff/passff#343]
### General information. Describe your environment

- Versions
  - Operating system: Arch Linux
  - Browser: Firefox 60.0.2 (64-bit)
  - PassFF: 1.3
  - Host app: 1.0.1

Status line output: `[10:52:12] ls -> (0)`

- Your preferences
  - In PassFF: not changed from the default
  - In the host app `passff.py`: not changed from the default
- Messages in the web console: none

---

### Actual behaviour
When opening a cached page that uses HTTP authentication, each parallel request to the server (images, javascript, etc.) spawns a new PassFF dialog, potentially spawning hundreds.

### Expected behaviour
PassFF should display at most one HTTP auth dialog at a time.

### Steps to reproduce the behaviour
1. Go to a cached web page with multiple images/javascript files that uses HTTP authentication
2. Observe that for each request, a PassFF dialog is spawned

For me, this is especially noticeable with a [Netdisco](http://netdisco.org/) install.

This issue might be related to #322.

**Comment by tuxor1337** [comment]
Can we somehow check whether two auth requests are exptected to be resolved by the same credentials?

**Comment by shtrom** [comment]
Sounds like #322

**Comment by arthaud** [comment]
I also have this problem (spawns 11 dialogs for a specific url). I would love to see a fix.

**Comment by shtrom** [comment]
Still happening here. It actually killed my Firefox with 500+ windows yesterday :laughing: 

As I said before, this seems to happen on cached pages, where FF tries to re-fetch all the objects at once, and gets separate 403 for each.

I'm wondering whether the number of concurrent popups could be limited by, say, one per combination of hostname and HTTP Auth Realm. This should do the trick to limit the requests for resources on the same server.

**Comment by shtrom** [comment]
Ok, I got a small hacking reproduction case for this. Essentially, it's just a small express-based server that asks for HTTP Auth on anything but the index. The index has two images in it.
[passff-auth.zip](https://github.com/passff/passff/files/6967978/passff-auth.zip)

When hitting http://localhost:3000, passff spawns one pass-entry window per image.
![Screenshot from 2021-08-11 21-28-07](https://user-images.githubusercontent.com/160280/129021345-38211db4-0776-4f5d-8dee-82f3a53c7a5e.png)

**Comment by tuxor1337** [comment]
Maybe we could just assume that all auth requests that come from one tab or frame or domain (or realm: https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/webRequest/onAuthRequired#details) can be resolved with the same credentials and only issue a single dialog window? In theory, I still don't know how PassFF should know that two auth requests can be handled together because they require the same credentials. But maybe in practice the tab, frame and domain of origin are already enough to decide whether two requests belong together.

**Comment by shtrom** [comment]
Yeah, that's how I'm trying to fix it: if the originUrl and the realm are the same as an existing popup, I skip the popup. I'll push a PR in a moment, but it's not ready yet.

**Comment by shtrom** [comment]
Wow, I'm seeing a weird behaviour that I can't quite explain. For context, I want to store the Promise in the auth, so any subsequent requests for the same Root URI/Real return the same Promise, rather than create a new one.

If I try to capture the Promise, in https://github.com/shtrom/passff/blob/726910c2f4c1912e13a0acc3e5d88b816246b23c/src/modules/auth.js#L79, _in addition to_ returning it,

```
diff --git a/src/modules/auth.js b/src/modules/auth.js
index 89ea740..26c356e 100644
--- a/src/modules/auth.js
+++ b/src/modules/auth.js
@@ -52,7 +52,7 @@ PassFF.Auth = (function () {
     auth.requestId = details.requestId;
     auth.requestUrl = details.url;
     auth.contextItems = PassFF.Pass.getUrlMatchingItems(auth.requestUrl);
-    return new Promise((resolve, reject) => {
+    return auth.promise = new Promise((resolve, reject) => {
       auth.resolve = resolve;
       PassFF.Page.goToAutoFillPending()
         .then(function (pending) {
```

things break, with the selection popup being empty, and the following errors in the log:
```
[PassFF.background] Message Auth.getAuthForPopup received in background util.js:66:14
[PassFF.itemPicker] Runtime port has crashed: Auth.getAuthForPopup <unavailable> util.js:66:14
    logPrototype moz-extension://fc230747-0377-431b-922e-41ca882599a5/modules/util.js:66
    background_exec moz-extension://fc230747-0377-431b-922e-41ca882599a5/modules/util.js:137
    (Async: promise callback)
    background_exec moz-extension://fc230747-0377-431b-922e-41ca882599a5/modules/util.js:136
    background_function moz-extension://fc230747-0377-431b-922e-41ca882599a5/modules/util.js:118
    restoreFromState moz-extension://fc230747-0377-431b-922e-41ca882599a5/modules/menu.js:59
    (Async: promise callback)
    restoreFromState moz-extension://fc230747-0377-431b-922e-41ca882599a5/modules/menu.js:59
    (Async: promise callback)
    init moz-extension://fc230747-0377-431b-922e-41ca882599a5/modules/menu.js:490
    init_promise moz-extension://fc230747-0377-431b-922e-41ca882599a5/modules/main.js:172
    (Async: promise callback)
    init moz-extension://fc230747-0377-431b-922e-41ca882599a5/modules/main.js:165
    (Async: EventListener.handleEvent)
    <anonymous> moz-extension://fc230747-0377-431b-922e-41ca882599a5/modules/main.js:199
TypeError: can't access property "requestUrl", auth is undefined menu.js:62:11
    restoreFromState moz-extension://fc230747-0377-431b-922e-41ca882599a5/modules/menu.js:62
    (Async: promise callback)
    restoreFromState moz-extension://fc230747-0377-431b-922e-41ca882599a5/modules/menu.js:60
    (Async: promise callback)
    init moz-extension://fc230747-0377-431b-922e-41ca882599a5/modules/menu.js:490
    init_promise moz-extension://fc230747-0377-431b-922e-41ca882599a5/modules/main.js:172
    (Async: promise callback)
    init moz-extension://fc230747-0377-431b-922e-41ca882599a5/modules/main.js:165
    (Async: EventListener.handleEvent)
    <anonymous> moz-extension://fc230747-0377-431b-922e-41ca882599a5/modules/main.js:199
```

If I navigate and select a password entry, I get those two new errors:
```
TypeError: can't access property "requestId", menuState.auth is undefined menu.js:569:7
    onPickItem moz-extension://fc230747-0377-431b-922e-41ca882599a5/modules/menu.js:569
    createItemMenuList moz-extension://fc230747-0377-431b-922e-41ca882599a5/modules/menu.js:320
    createMenuList moz-extension://fc230747-0377-431b-922e-41ca882599a5/modules/menu.js:274
    onListItemSelected moz-extension://fc230747-0377-431b-922e-41ca882599a5/modules/menu.js:197
    (Async: EventListener.handleEvent)
    createMenuItem moz-extension://fc230747-0377-431b-922e-41ca882599a5/modules/menu.js:372
    createItemsMenuList moz-extension://fc230747-0377-431b-922e-41ca882599a5/modules/menu.js:310
    forEach self-hosted:205
    createItemsMenuList moz-extension://fc230747-0377-431b-922e-41ca882599a5/modules/menu.js:292
    createMenuList moz-extension://fc230747-0377-431b-922e-41ca882599a5/modules/menu.js:272
    onSearchKeyup moz-extension://fc230747-0377-431b-922e-41ca882599a5/modules/menu.js:142
    (Async: EventListener.handleEvent)
    init_ui moz-extension://fc230747-0377-431b-922e-41ca882599a5/modules/menu.js:432
    init moz-extension://fc230747-0377-431b-922e-41ca882599a5/modules/menu.js:489
    init_promise moz-extension://fc230747-0377-431b-922e-41ca882599a5/modules/main.js:172
    (Async: promise callback)
    init moz-extension://fc230747-0377-431b-922e-41ca882599a5/modules/main.js:165
    (Async: EventListener.handleEvent)
    <anonymous> moz-extension://fc230747-0377-431b-922e-41ca882599a5/modules/main.js:199
TypeError: can't access property "requestId", menuState.auth is undefined menu.js:569:7
    onPickItem moz-extension://fc230747-0377-431b-922e-41ca882599a5/modules/menu.js:569
    createItemMenuList moz-extension://fc230747-0377-431b-922e-41ca882599a5/modules/menu.js:320
    createMenuList moz-extension://fc230747-0377-431b-922e-41ca882599a5/modules/menu.js:274
    onListItemSelected moz-extension://fc230747-0377-431b-922e-41ca882599a5/modules/menu.js:197
    (Async: EventListener.handleEvent)
    createMenuItem moz-extension://fc230747-0377-431b-922e-41ca882599a5/modules/menu.js:372
    createItemsMenuList moz-extension://fc230747-0377-431b-922e-41ca882599a5/modules/menu.js:310
    forEach self-hosted:205
    createItemsMenuList moz-extension://fc230747-0377-431b-922e-41ca882599a5/modules/menu.js:292
    createMenuList moz-extension://fc230747-0377-431b-922e-41ca882599a5/modules/menu.js:272
    onSearchKeyup moz-extension://fc230747-0377-431b-922e-41ca882599a5/modules/menu.js:142
    (Async: EventListener.handleEvent)
    init_ui moz-extension://fc230747-0377-431b-922e-41ca882599a5/modules/menu.js:432
    init moz-extension://fc230747-0377-431b-922e-41ca882599a5/modules/menu.js:489
    init_promise moz-extension://fc230747-0377-431b-922e-41ca882599a5/modules/main.js:172
    (Async: promise callback)
    init moz-extension://fc230747-0377-431b-922e-41ca882599a5/modules/main.js:165
    (Async: EventListener.handleEvent)
    <anonymous> moz-extension://fc230747-0377-431b-922e-41ca882599a5/modules/main.js:199
```
This seems to happen any time I try to store the Promise into the `auth` object, under any name, and even if I do it in multiple steps:

``` let p = new Promise(...);
//auth.<whatever> = p; // if uncommented, this breaks as described above
return p
```

@tuxor1337 would you have an idea about what might be causing this? I'm not sure it's specific to PassFF, and may just be a subtlety of JS Promises I'm not aware of (note that the `auth` object that I want to store the promise in is also manipulated by the promise itself).

**Comment by shtrom** [comment]
Ok, I have sidestepped the problem by caching the promises in a separate array. @14mRh4X0r @arthaud (as you have commented here before) would you mind giving a spin to my PR #489, and see if it fixes the issues you were seeing?

**Comment by shtrom** [comment]
>  I still don't know how PassFF should know that two auth requests can be handled together (using the same credentials). 

@tuxor1337 I happened upon this https://datatracker.ietf.org/doc/html/rfc7235#section-2.2

> The "realm" authentication parameter is reserved for use by authentication schemes that wish to indicate a scope of protection.
> A protection space is defined by the canonical root URI (the scheme and authority components of the effective request URI; see Section 5.5 of [RFC7230]) of the server being accessed, in combination with the realm value if present.  These realms allow the protected resources on a server to be partitioned into a set of protection spaces, each with its own authentication scheme and/or authorization database.

So I look up existing `auth`s base on Root URI and Realm. I did add a fallback, though, where if the previous promise has resolved, but we still get a 401 (say, for a different object), we re-display the popup. This is in case of poor implementations that use the same Realm, but different credentials for different object within the same protection space.

**Comment by tuxor1337** [comment]
@shtrom Thanks a lot for your PR! I will probably have time to have a look, give feedback (and maybe merge) in 2 days or so.

**Comment by tuxor1337** [comment]
Thanks for the PR, it's merged now!

> Wow, I'm seeing a weird behaviour that I can't quite explain. For context, I want to store the Promise in the auth, so any subsequent requests for the same Root URI/Real return the same Promise, rather than create a new one.
>  ...
> @tuxor1337 would you have an idea about what might be causing this? I'm not sure it's specific to PassFF, and may just be a subtlety of JS Promises I'm not aware of (note that the `auth` object that I want to store the promise in is also manipulated by the promise itself).

This is because the `auth` object is generated by the background worker and sent later to the item picker pop-up. For that, the object needs to be serializable. The promise can't be serialized. I now remove the promise before sending the object:
https://github.com/passff/passff/blob/b2ae28553c2269fc662676588b3dfc500821a734/src/modules/auth.js#L148-L151

**Comment by shtrom** [comment]
Awesome! Thanks. Do you have a sense of when the next release will happen?

**Comment by tuxor1337** [comment]
@shtrom This fix is included in [1.11.3](https://github.com/passff/passff/releases/tag/1.11.3) which will be available from addons.mozilla.org very soon.
